Whats new in AOCL 3.2

amd-toolchain-support opened the pull request:
1) amdblis
    Progress feature for xGEMM and xTRSM APIs:
    Time taken to complete the mathematical operations tends to increase
    exponentially with large input problem sizes; this feature provides
    users a periodic update on the operation progress.
    Improvements in SGEMM, DGEMM, ZGEMM, DGEMV, Level-1 and Level-2
    routines
2) amdlibflame
    Improved SVD and Eigen Value routines
    New feature to track the computation progress of selected APIs
3) amdfftw
    Dynamic dispatcher for AOCL-FFTW
    Upgraded AOCL-FFTW to align with the reference FFTW 3.3.10
4) amdlibm
    New complex number variant functions – Most frequently used
    functions exp/log/pow
    Fast variants (with reduced precision – ULP up to 4) of the routines
    acos, asin, asinf, atan, atanf, expf, log, logf, powf, tan, and tanf
5) aocl-parse
    Two new APIs: One of them to multiply two sparse matrices and other
    an ILU based preconditioner API Sparsity pattern analysis framework for SPMV
6) amdscalapack
    New feature to track the computation progress of selected APIs
7) amd-aocl
    Bundle package for above libraries

amd-toolchain-support committed:
Whats new in AOCL 3.2

1) amdblis
    Progress feature for xGEMM and xTRSM APIs:
    Time taken to complete the mathematical operations tends to increase
    exponentially with large input problem sizes; this feature provides
    users a periodic update on the operation progress.
    Improvements in SGEMM, DGEMM, ZGEMM, DGEMV, Level-1 and Level-2
    routines
2) amdlibflame
    Improved SVD and Eigen Value routines
    New feature to track the computation progress of selected APIs
3) amdfftw
    Dynamic dispatcher for AOCL-FFTW
    Upgraded AOCL-FFTW to align with the reference FFTW 3.3.10
4) amdlibm
    New complex number variant functions – Most frequently used
    functions exp/log/pow
    Fast variants (with reduced precision – ULP up to 4) of the routines
    acos, asin, asinf, atan, atanf, expf, log, logf, powf, tan, and tanf
5) aocl-parse
    Two new APIs: One of them to multiply two sparse matrices and other
    an ILU based preconditioner API Sparsity pattern analysis framework for SPMV
6) amdscalapack
    New feature to track the computation progress of selected APIs
7) amd-aocl
    Bundle package for above libraries

tldahlgren commented on var/spack/repos/builtin/packages/amd-aocl/package.py:
```diff
@@ -0,0 +1,36 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
```
```suggestion
# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
```

amd-toolchain-support replied:
Done

tldahlgren reviewed (commented):
Confirmed sha256.  The copyright for `amd-aocl` needs to be corrected.

alalazo commented on var/spack/repos/builtin/packages/amdblis/package.py:
```diff
@@ -23,13 +23,19 @@ class Amdblis(BlisBase):
+    conflicts(
+        '+ilp64',
+        when='@:3.0.0',
+        msg='ilp64 is supported from amdblis 3.0.1 version onwards')
```
A conflict is not needed since the variant is conditional to `@3.0.1:`

amd-toolchain-support replied:
Changes incorporated

alalazo commented on var/spack/repos/builtin/packages/amdfftw/package.py:
```diff
@@ -70,6 +71,11 @@ class Amdfftw(FftwBase):
         'amd-app-opt',
         default=False,
         description='Build with amd-app-opt suppport')
+    variant(
+        'amd-dynamic-dispatcher',
+        default=False,
+        description='Single portable optimized library'
```
```suggestion
    variant(
        'amd-dynamic-dispatcher',
        default=False,
        when='@3.2:',
        description='Single portable optimized library'
```

amd-toolchain-support replied:
Changes incorporated

alalazo commented on var/spack/repos/builtin/packages/amdfftw/package.py:
```diff
@@ -175,6 +177,14 @@ class Amdfftw(FftwBase):
+        '+amd-dynamic-dispatcher',
+        when='%aocc',
+        msg='dynamic-dispatcher is not supported by AOCC clang compiler')
+    conflicts(
+        '+amd-dynamic-dispatcher',
+        when='@:3.1',
+        msg='dynamic-dispatcher is supported from 3.2 version onwards')
```
This conflict can be removed making the variant conditional.

amd-toolchain-support replied:
Changes incorporated

tldahlgren commented on var/spack/repos/builtin/packages/amdscalapack/package.py:
```diff
@@ -37,7 +38,9 @@ class Amdscalapack(ScalapackBase):
 
     def url_for_version(self, version):
-        if version == Version('3.1'):
+        if version == Version('3.2'):
+            return "https://github.com/amd/aocl-scalapack/archive/3.2.tar.gz"
+        elif version == Version('3.1'):
```
I agree somewhat with #31521 about the version handling here needing to be changed.  For example, this package could use something along the following for the body of this method:

```
        vers = "https://github.com/amd/{0}/archive/{1}.tar.gz"
        if version >= Version('3.1'):
            return vers.format('aocl-scalapack', version)

        return vers.format('scalapack', version)
```

amd-toolchain-support replied:
Changes incorporated

amd-toolchain-support committed:
Incorporate reviewers comments

amd-toolchain-support commented:
> `amd-aocl ~openmp` doesn't depend on non-openmp variants and might end up using openmp-enabled installed dependencies.

How to handle this situation, please guide us.

zzzoom commented on var/spack/repos/builtin/packages/amd-aocl/package.py:
```diff
@@ -0,0 +1,36 @@
+
+    for vers in ['2.2', '3.0', '3.1', '3.2']:
+        depends_on('amdblis@{0} threads=openmp'.format(vers), when='@{0} +openmp'.format(vers))
+        depends_on('amdblis@{0}'.format(vers), when='@{0} ~openmp'.format(vers))
```
`depends_on('amdblis@{0} threads=none'.format(vers), when='@{0} ~openmp'.format(vers))`

amd-toolchain-support replied:
Changes incorporated

zzzoom commented on var/spack/repos/builtin/packages/amd-aocl/package.py:
```diff
@@ -0,0 +1,36 @@
+        depends_on('amdblis@{0} threads=openmp'.format(vers), when='@{0} +openmp'.format(vers))
+        depends_on('amdblis@{0}'.format(vers), when='@{0} ~openmp'.format(vers))
+        depends_on('amdfftw@{0} +openmp'.format(vers), when='@{0} +openmp'.format(vers))
+        depends_on('amdfftw@{0}'.format(vers), when='@{0} ~openmp'.format(vers))
```
`depends_on('amdfftw@{0} ~openmp'.format(vers), when='@{0} ~openmp'.format(vers))`

amd-toolchain-support replied:
Changes incorporated

amd-toolchain-support committed:
Changes to fix following problem:

amd-aocl ~openmp doesn't depend on non-openmp variants and might end up using openmp-enabled installed dependencies.

alalazo commented on var/spack/repos/builtin/packages/amdlibflame/package.py:
```diff
@@ -69,21 +70,32 @@ def lapack_libs(self):
-        complex types when compiling with aocc flang"""
-        if "@3.0: %aocc" in self.spec:
-            args.append("--enable-f2c-dotc")
+        # 3.2.rc2 version onwards amd optimized flags are encapsulated under:
+        # enable-amd-flags for gcc compiler
+        # enable-amd-aocc-flags for aocc compiler
+        if "@3.2.rc2:" in self.spec:
+            if "%gcc" in self.spec:
+                args.append("--enable-amd-flags")
+            if "%aocc" in self.spec:
+                args.append("--enable-amd-aocc-flags")
```
I don't think this portion of the code will apply to any version:
```console
Spack version 0.18.0
Python 3.8.10, Linux x86_64
>>> import spack.version
>>> v1 = spack.version.ver('3.2')
>>> v2 = spack.version.ver('3.2.rc2')
>>> v1 < v2
True
```

amd-toolchain-support replied:
Thanks for notifying the issue. The issue is now corrected.

This was meant for internal test purposes. 
My bad, we missed updating the amdlibflame recipe for the final release version.

alalazo commented on var/spack/repos/builtin/packages/amdlibflame/package.py:
```diff
@@ -69,21 +70,32 @@ def lapack_libs(self):
 
-        if "@3.0.1: +ilp64" in self.spec:
-            args.append("--enable-ilp64")
+        if "@:3.1" in self.spec:
```
Here and below there are a lot of overlapping conditions. I just wanted to double check that this was on purpose.

amd-toolchain-support replied:
Yes. 

For example, configure option for 'FORTRAN to C calling convention for  complex types when compiling with aocc flang' was:

- For version 3.0:3.1 it is  "--enable-f2c-dotc" 
- From version 3.1 onward it is "--enable-void-return-complex"

alalazo reviewed (changes requested):
Mostly LGTM. Only two minor comments

amd-toolchain-support committed:
change to fix wrong version reference
